src/data/setoid/partition.lean

```diff
@@ -201,3 +212,134 @@ _ (subtype (@is_partition α)) _ (partial_order.to_preorder _) $ partition.rel_i
 end partition
 
 end setoid
+
+/-- A partition of a type indexed by another type. -/
+@[nolint has_inhabited_instance]
+structure indexed_partition {ι α : Type*} (s : ι → set α) :=
+(eq_of_mem : ∀ {x i j}, x ∈ s i → x ∈ s j → i = j)
+(some : ι → α)
+(some_mem : ∀ i, some i ∈ s i)
+(index : α → ι)
+(mem_index : ∀ x, x ∈ s (index x))
+
+/-- The non-constructive constructor for `indexed_partition`. -/
+noncomputable
+def indexed_partition.mk' {ι α : Type*} (s : ι → set α) (dis : ∀ i j, i ≠ j → disjoint (s i) (s j))
+(empty : ∀ i, (s i).nonempty) (ex : ∀ x, ∃ i, x ∈ s i) : indexed_partition s :=
+{ eq_of_mem := begin
+    classical,
+    intros x i j hxi hxj,
+    by_contra h,
+    exact dis _ _ h ⟨hxi, hxj⟩
+  end,
+  some := λ i, (empty i).some,
+  some_mem := λ i, (empty i).some_spec,
+  index := λ x, (ex x).some,
+  mem_index := λ x, (ex x).some_spec }
+
+namespace indexed_partition
+
+open set
+
+variables {ι α : Type*} {s : ι → set α} (hs : indexed_partition s)
+include hs
+
+lemma exists_mem (x : α) : ∃ i, x ∈ s i := ⟨hs.index x, hs.mem_index x⟩
+
+lemma Union : (⋃ i, s i) = univ :=
+by { ext x, simp [hs.exists_mem x] }
+
+lemma disjoint : ∀ {i j}, i ≠ j → disjoint (s i) (s j) :=
+λ i j h x ⟨hxi, hxj⟩,h (hs.eq_of_mem hxi hxj)
+
+lemma eq_index_of_mem {x i} (hxi : x ∈ s i) : i = hs.index x :=
+hs.eq_of_mem hxi (hs.mem_index x)
+
+/-- The equivalence relation associated to an indexed partition. Two
+elements are equivalent if they belong to the same set of the partition. -/
+protected def setoid (hs : indexed_partition s) : setoid α :=
+{ r := λ x y, ∃ i, x ∈ s i ∧ y ∈ s i,
```

Comment:
I think this golfs to `setoid.ker hs.index`?

Reply:
This won't be definitionaly equal, so all proofs will break. I don't have time to play this game.

Reply:
I think it's better here to use `setoid.ker` in this file - we have it for a reason, and it feels unnecessary to duplicate work when there's a nicer definition which also simplifies later proofs as Eric shows below.

Reply:
You are the ones who talk about duplicating work. That file works, it is usable since it has been used without any problem. Now you want to duplicate work by redoing it differently. You are welcome do it, but I won't.

Reply:
I think it's important to distinguish duplicating effort from duplicating code. Bhavik means that the API surface of mathlib has duplication (aka, the output of the work is duplicated), while Patrick is objecting to redoing the work.

> You are welcome do it, but I won't.

I've pushed a change to that effect, which removes 20 lines of proofs.